### ISSUE TITLE
How to triangulate point from multiple views?

### ISSUE THREAD

**Issue by antonaaa** [colmap/colmap#384]
I saw these code in file colmap/src/base/triangulation.cc.
I want to know the mathematics behind these codes, could you tell me about that or any materials suggest?

![_062](https://user-images.githubusercontent.com/20875134/41403109-a3c880ea-6ff6-11e8-898c-1f584fa5c174.png)

**Comment by IdoMa** [comment]
Thanks @Linusnie for resurfacing this question - I also found this confusing (and couldn't find references). 

It's clear why this residual vanishes when the re-projection distance vanishes. To me though it's unclear what it expresses when it doesn't.

Would be happy for any explanations and\or references (@ahojnnes).

**Comment by IdoMa** [comment]
Digging in some more, it seems like the sum of squared angular errors (=between back-projected and re-projected rays) is being minimized.

![Screenshot 2023-06-22 at 18 17 52](https://github.com/colmap/colmap/assets/35418794/a4408b80-7dc1-4429-8336-0c38fbe7e496)

![Screenshot 2023-06-22 at 18 21 25](https://github.com/colmap/colmap/assets/35418794/12a25eba-803a-42d6-ad62-e3045d47a586)

Worryingly, the objective seems to depend on the **norm** of the re-projections (which in the context of homogenous coordinates should be meaningless).

Is this derivation and analysis correct?

**Comment by Linusnie** [comment]
@IdoMa you're right that the objective doesn't correspond to anything meaningful geometrically, but as long as the observations noise is reasonable it won't make a big difference in the estimated point (see graph 1+2 in the paper). The point will also be further refined with reprojection errors in bundle adjustment

**Comment by yanming-zou** [comment]
I have been investigating this problem recently. COLMAP's implementation uses a mid-point method. For more details, please refer to Formula 11.24 in Section 11.2.4 of the textbook *Computer Vision: Algorithms and Applications, 2nd Edition* by R. Szeliski. COLMAP's implementation differs from the textbook in that it uses pixel values instead of normalized coordinates. Thus, a projection matrix is needed.
